### DISCUSSION TITLE
Automatic Ouraring health metrics sync to Obsidian

### DISCUSSION THREAD

**Discussion by louis030195** [SilentVoid13/Templater#971] [Templates Showcase]
```bash
curl -X GET 'https://api.ouraring.com/v2/usercollection/daily_readiness?start_date=$(date +"%Y-%m-%d")&end_date=$(date +"%Y-%m-%d")' \
    -H "Authorization : Bearer $TOKEN" \
    -H "Accept: application/json" \
    -H "Content-Type: application/json" | /opt/homebrew/bin/jq '.data[0].score'

curl -X GET 'https://api.ouraring.com/v2/usercollection/daily_sleep?start_date=$(date +"%Y-%m-%d")&end_date=$(date +"%Y-%m-%d")' \
    -H "Authorization : Bearer $TOKEN" \
    -H "Accept: application/json" \
    -H "Content-Type: application/json" | /opt/homebrew/bin/jq '.data[0].score'


curl -X GET 'https://api.ouraring.com/v2/usercollection/daily_activity?start_date=$(date +"%Y-%m-%d")&end_date=$(date +"%Y-%m-%d")' \
    -H "Authorization : Bearer $TOKEN" \
    -H "Accept: application/json" \
    -H "Content-Type: application/json" | /opt/homebrew/bin/jq '.data[0].score'
```

```md
Ouraring readiness: <% tp.user.ouraring_readiness() %>
Ouraring sleep: <% tp.user.ouraring_sleep() %>
Ouraring activity: <% tp.user.ouraring_activity() %>
```

**Comment by SimpleMedia1** [comment]
Hey @louis030195, I wanted to do this as well with my oura ring. Could you expand on the steps, or create a short video tutorial please.

**Comment by kadenzipfel** [comment]
as a note, wrapping the URL in double quotes seems to be necessary on OSX

**Comment by steven-serka** [comment]
I saw this when I first started using Obsidian and Templater, so thank you for the post. 

I expanded upon your work and idea, creating a specific script file and a template to utilize it.

edit: the following script is meant to be use in the daily note template hence the reference to `tp.file.title` to generate date

script: generate_ouraring_score.js
```javascript
async function generate_ouraring_score(tp, stat) {
    const fmt = "YYYY-MM-DD";
    const now = tp.date.now(fmt, 0, tp.file.title, fmt);

    const baseUrl = "https://api.ouraring.com/v2/usercollection/";
    const fullUrl = baseUrl + stat + '?start_date=' + now + '&end_date=' + now;

    const response = JSON.parse(
        await request({
            url: fullUrl,
            method: "GET",
            contentType: "application/json",
            headers: {
                'Authorization': 'Bearer TOKEN',
                'Content-Type': 'application/json',
                'Accept': 'application/json'
            },
        })
    );
    
// only return a value if the response contains it
    if(response && response.data && response.data.length > 0 && response.data[0].score)
        return response.data[0].score
    
    return 'N/A';
}

module.exports = generate_ouraring_score;
```

Templater template: 
```obsidian
# Oura Ring #oura
<!-- Oura Ring Stats -->
**Readiness**: *<% tp.user.generate_ouraring_score(tp, "daily_readiness") %>*
**Sleep**: *<% tp.user.generate_ouraring_score(tp, "daily_sleep") %>*
**Activity**: *<% tp.user.generate_ouraring_score(tp, "daily_activity") %>
```

Final output:
<img width="265" alt="image" src="https://github.com/SilentVoid13/Templater/assets/31013786/8123d870-3551-4f58-aeda-52ac37b974f4">

Note: You will need to generate a personal access token on the [oura ring cloud developer api dashboard](https://cloud.ouraring.com/personal-access-tokens) and insert it into the javascript file where TOKEN is shown. Completely replace TOKEN with the generated token value. This personal access token only allows you to get data from YOUR account and nobody else's.
Example:  
`'Authorization': 'Bearer AWQ2WEQF2YSJDDUV7COJCHHXWASAW5F4'`


If you want to expand further beyond just the `/daily_readiness`, `/daily_sleep`, and `daily_activity` endpoints. Consider creating a new script or adjusting the one above to use the corresponding endpoints found in the [oura ring API documentation](https://cloud.ouraring.com/v2/docs#section/Overview)

**Reply by jonngyu** [reply]
Hi,

I've implemented your templates to get the data from my Oura ring, but it seems like only Readiness and Sleep scores get pulled. Have you had any luck pulling the Activity Score? Cause it appears it didn't even pull for you as well?

**Reply by steven-serka** [reply]
Hi, I've recently realized this as well and have a backlog item to look into this soon

**Comment by SimpleMedia1** [comment]
Curious if any of you have integrated the api with Notion as well?
